### ISSUE TITLE
flow_from_directory counter issue. Requires manual reset.

### ISSUE THREAD

**Issue by don-tpanic** [keras-team/keras-preprocessing#196]
Code to replicate my findings: (Notice you will have to replace `imagenet_test` with your own image directory)
```
import numpy as np
import keras
from keras.preprocessing.image import ImageDataGenerator
from keras.applications.vgg16 import preprocess_input
from keras.optimizers import Adam
from keras.applications import VGG16
from keras import metrics
from keras import backend as K
K.tensorflow_backend._get_available_gpus()


model = VGG16(weights='imagenet', include_top=True, input_shape=(224, 224, 3))
model.compile(Adam(lr=0.001),
                 loss='sparse_categorical_crossentropy',
                 metrics=['accuracy'])

batch_size = 250
imagenet_test = '/mnt/fast-data17/datasets/ILSVRC/2012/clsloc/val/'
datagen_test = ImageDataGenerator(preprocessing_function=preprocess_input)

test_generator = datagen_test.flow_from_directory(
                                            imagenet_test,
                                            batch_size=batch_size,
                                            shuffle=False,
                                            class_mode="sparse",
                                            target_size=(224, 224)
                                            )


steps = len(test_generator.classes) / batch_size
print('batch_index1: ', test_generator.batch_index)
check1 = model.predict_generator(test_generator, steps=steps, verbose=1)
print('batch_index2: ', test_generator.batch_index)
check2 = model.predict_generator(test_generator, steps=steps, verbose=1)

print(check1[0, :][0])
print(check2[0, :][0])
```
=========================================================================
It seems by running predict_generator twice would yield different results. I looked into this and found that the counter doesn't reset when the generator is done. I believe the problem is in `keras-preprocessing/keras_preprocessing/image/iterator.py`
More specifically, it's in the `def _flow_index(self)` function:
```
class Iterator(IteratorType):
    """Base class for image data iterators.
    Every `Iterator` must implement the `_get_batches_of_transformed_samples`
    method.
    # Arguments
        n: Integer, total number of samples in the dataset to loop over.
        batch_size: Integer, size of a batch.
        shuffle: Boolean, whether to shuffle the data between epochs.
        seed: Random seeding for data shuffling.
    """
    white_list_formats = ('png', 'jpg', 'jpeg', 'bmp', 'ppm', 'tif', 'tiff')

    def __init__(self, n, batch_size, shuffle, seed):
        self.n = n
        self.batch_size = batch_size
        self.seed = seed
        self.shuffle = shuffle
        self.batch_index = 0
        self.total_batches_seen = 0
        self.lock = threading.Lock()
        self.index_array = None
        self.index_generator = self._flow_index()

    def _set_index_array(self):
        self.index_array = np.arange(self.n)
        if self.shuffle:
            self.index_array = np.random.permutation(self.n)

    def __getitem__(self, idx):
        if idx >= len(self):
            raise ValueError('Asked to retrieve element {idx}, '
                             'but the Sequence '
                             'has length {length}'.format(idx=idx,
                                                          length=len(self)))
        if self.seed is not None:
            np.random.seed(self.seed + self.total_batches_seen)
        self.total_batches_seen += 1
        if self.index_array is None:
            self._set_index_array()
        index_array = self.index_array[self.batch_size * idx:
                                       self.batch_size * (idx + 1)]
        return self._get_batches_of_transformed_samples(index_array)

    def __len__(self):
        return (self.n + self.batch_size - 1) // self.batch_size  # round up

    def on_epoch_end(self):
        self._set_index_array()

    def reset(self):
        self.batch_index = 0

    def _flow_index(self):
        # Ensure self.batch_index is 0.
        self.reset()
        while 1:
            if self.seed is not None:
                np.random.seed(self.seed + self.total_batches_seen)
            if self.batch_index == 0:
                self._set_index_array()

            current_index = (self.batch_index * self.batch_size) % self.n
            print('batch_index = ', self.batch_index, 'current_index = ', current_index)
            if self.n > current_index + self.batch_size:
                self.batch_index += 1
            else:
                self.batch_index = 0
            self.total_batches_seen += 1
            yield self.index_array[current_index:
                                   current_index + self.batch_size]
```
Notice I add this `print('batch_index = ', self.batch_index, 'current_index = ', current_index)` after `current_index = (self.batch_index * self.batch_size) % self.n` in the code above, and here are the results upon the first `predict_generator` has terminates:
```
batch_index =  0 current_index =  0
batch_index =  1 current_index =  250
....
batch_index =  197 current_index =  49250
batch_index =  198 current_index =  49500
batch_index =  199 current_index =  49750
batch_index =  0 current_index =  0
``` 
This explains why running two `predict_generator` gives different results. It's because the `batch_index` would start from `1` when executing the second `predict_generator`. Hence in cases like this, my current work-round is call `test_generator.reset()` before the second `predict_generator` .

**Comment by Vijayabhaskar96** [comment]
This is a known behavior, but I do think it should reset automatically on every call of `predict_generator`, Maybe an attribute like `reset_generator=True` with True as its default value, that resets the `test_generator`? @rragundez What do you think?

**Comment by rragundez** [comment]
As @Vijayabhaskar96 mentions, the generator will not be reseted, and this is the desirable behavior (because it is a generator, if not you enter the rabbit hole of that does being "done" means in a generator context). In the case you want this to happen that's why indeed the `reset` method exists. No need for an extra argument. 
I am confused by you issue description @don-tpanic, if your batch size times number of steps is equal to the number of observations, you will get the same output from the generator every time you ran those amount of steps (given that shuffle is False of course).

here a minimal reproducible example:
```
from keras_preprocessing.image import ImageDataGenerator
import numpy as np

N = 4
batch_size = 2
steps = N // batch_size

data_array = np.random.randint(0, 10, (N, 1, 1, 1))
data_gen = ImageDataGenerator().flow(data_array, batch_size=batch_size, shuffle=False)
for i in range(steps):
    print("\nBatch index:", data_gen.batch_index, "\n")
    print(data_gen.next())
    
print("\n\nTIGGER AGAIN\n\n")
for i in range(steps):
    print("Batch index:", data_gen.batch_index)
    print(data_gen.next())
>>>
Batch index: 0 

[[[[1.]]]


 [[[9.]]]]

Batch index: 1 

[[[[3.]]]


 [[[6.]]]]


TIGGER AGAIN


Batch index: 0
[[[[1.]]]


 [[[9.]]]]
Batch index: 1
[[[[3.]]]


 [[[6.]]]]
```
As you can see you get exactly the same output from both for loops.

**Comment by rragundez** [comment]
@don-tpanic  is it clearer now? or can I help you clarify something else? I will close the ticket for now.

**Comment by epignatelli** [comment]
Even using manual reset `generator.reset()`, `total_batches_seen` doesn't roll back to 0.
Is this expected as well?

**Comment by don-tpanic** [comment]
> Even using manual reset `generator.reset()`, `total_batches_seen` doesn't roll back to 0.
> Is this expected as well?

I think `generator.reset()` only rests `batch_index` to 0

**Comment by don-tpanic** [comment]
@rragundez Apologies for such a slow response. I have run the example you provided but somehow I am getting a different behaviour for `Batch index`
`
Batch index: 0 

[[[[6.]]]


 [[[3.]]]]

Batch index: 1 

[[[[7.]]]


 [[[4.]]]]


TIGGER AGAIN



Batch index: 0 

[[[[6.]]]


 [[[3.]]]]

Batch index: 1 

[[[[7.]]]


 [[[4.]]]]
`

**Comment by rragundez** [comment]
That's indeed strange, @Vijayabhaskar96  do you mind running the minimal reproducible example I posted and sharing the output?

**Comment by Vijayabhaskar96** [comment]
@rragundez 
Here you go!

```
Batch index: 0

[[[[4.]]]


 [[[8.]]]]

Batch index: 1

[[[[5.]]]


 [[[0.]]]]


TIGGER AGAIN



Batch index: 0

[[[[4.]]]


 [[[8.]]]]

Batch index: 1

[[[[5.]]]


 [[[0.]]]]
```

**Comment by rragundez** [comment]
thanks! @Vijayabhaskar96 , oh well.. @don-tpanic

**Comment by don-tpanic** [comment]
> thanks! @Vijayabhaskar96 , oh well.. @don-tpanic

Ha..I don't know what's going on there..perhaps proof of parallel realities
